rtpreceiver.go

```diff
@@ -14,149 +13,190 @@ type RTPReceiver struct {
-				return
-			}
-
-			var rtpPacket rtp.Packet
-			if err = rtpPacket.Unmarshal(append([]byte{}, readBuf[:rtpLen]...)); err != nil {
-				pcLog.Warnf("Failed to unmarshal RTP packet, discarding: %v \n", err)
-				continue
-			}
-
-			if !payloadSet {
-				r.Track.PayloadType = rtpPacket.PayloadType
-				payloadSet = true
-				close(r.hasRecv)
+				r.rtcpReadErr <- err
+				break
 			}
 
 			select {
-			case r.rtpOut <- &rtpPacket:
-			default:
+			case in := <-r.rtcpReadBuf:
+				copy(in, readBuf[:i])
+				r.rtcpReadLen <- i
+			default: // Discard if we have no inbound read
 			}
 		}
 	}()
 
-	// RTCP ReadLoop
 	go func() {
-		defer func() {
-			close(r.rtcpOut)
-			close(r.rtcpOutDone)
-		}()
-
-		srtcpSession, err := r.transport.getSRTCPSession()
-		if err != nil {
-			pcLog.Warnf("Failed to open SRTCPSession, Track done for: %v %d \n", err, parameters.encodings.SSRC)
-			return
-		}
-
-		readStream, err := srtcpSession.OpenReadStream(parameters.encodings.SSRC)
-		if err != nil {
-			pcLog.Warnf("Failed to open RTCP ReadStream, Track done for: %v %d \n", err, parameters.encodings.SSRC)
-			return
-		}
-		r.mu.Lock()
-		r.rtcpReadStream = readStream
-		r.mu.Unlock()
-
 		readBuf := make([]byte, receiveMTU)
 		for {
-			rtcpLen, err := readStream.Read(readBuf)
+			i, err := r.rtpReadStream.Read(readBuf)
 			if err != nil {
-				pcLog.Warnf("Failed to read, Track done for: %v %d \n", err, parameters.encodings.SSRC)
-				return
+				r.rtpReadErr <- err
+				break
 			}
 
-			rtcpPacket, _, err := rtcp.Unmarshal(append([]byte{}, readBuf[:rtcpLen]...))
-			if err != nil {
-				pcLog.Warnf("Failed to unmarshal RTCP packet, discarding: %v \n", err)
-				continue
-			}
 			select {
-			case r.rtcpOut <- rtcpPacket:
-			default:
+			case in := <-r.rtpReadBuf:
+				copy(in, readBuf[:i])
+				r.rtpReadLen <- i
+			default: // Discard if we have no inbound read
 			}
 		}
 	}()
 
-	return r.hasRecv
+	return nil
+}
+
+// Read reads incoming RTCP for this RTPReceiver
+func (r *RTPReceiver) Read(b []byte) (n int, err error) {
+	if initialReadErr := func() error {
+		r.mu.RLock()
+		defer r.mu.RUnlock()
+		switch {
+		case !r.received:
+			return fmt.Errorf("RTPReceiver was never started")
+		case r.closed:
+			return fmt.Errorf("RTPReceiver is closed")
+		case r.rtcpLastReadErr != nil:
+			return r.rtcpLastReadErr
+		default:
+			return nil
+		}
+	}(); initialReadErr != nil {
+		return 0, initialReadErr
+	}
+
+	select {
+	case err := <-r.rtcpReadErr:
+		r.mu.Lock()
+		r.rtcpLastReadErr = err
+		r.mu.Unlock()
+		return 0, err
+	default: // noop
+	}
+
+	select {
+	case r.rtcpReadBuf <- b:
+	case <-r.chansDone:
+		return 0, fmt.Errorf("RTPReceiver is stopped")
+	}
+
+	select {
+	case i := <-r.rtcpReadLen:
+		return i, nil
+	case <-r.chansDone:
+		return 0, fmt.Errorf("RTPReceiver is stopped")
+	}
+}
+
+// ReadRTCP is a convenience method that wraps Read and unmarshals for you
+func (r *RTPReceiver) ReadRTCP(b []byte) (rtcp.Packet, rtcp.Header, error) {
```

Comment:
`rtcp.Packet` is an interface that includes `Header()` so we don't need to split out like this anymore. You can just return the packet.

Reply:
YAY